File: gr-starcoder/python/qa_enqueue_message_sink.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# 
# Copyright 2018 Infostellar, Inc.
# 
# This is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3, or (at your option)
# any later version.
# 
# This software is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this software; see the file COPYING.  If not, write to
# the Free Software Foundation, Inc., 51 Franklin Street,
# Boston, MA 02110-1301, USA.
# 

from gnuradio import gr, gr_unittest
from gnuradio import blocks
import starcoder_swig as starcoder
import pmt
import time

class qa_enqueue_message_sink (gr_unittest.TestCase):

    def setUp (self):
        self.tb = gr.top_block ()

    def tearDown (self):
        self.tb = None

    def test_001_t (self):
        # The following source block is needed even if we don't use its message.
        # This is because the GNURadio scheduler will not create a thread for
        # blocks that aren't connected to anything. Thus the following block is
        # needed to have enqueue_message_sink connected to something.
        src = blocks.message_strobe(pmt.intern('unused message'), 99999)
        op = starcoder.enqueue_message_sink()

        msg1 = pmt.make_u8vector(3, ord('A'))
        msg2 = pmt.make_u8vector(20, ord('b'))

        op.to_basic_block()._post(pmt.intern('in'), msg1)
        op.to_basic_block()._post(pmt.intern('in'), msg2)

        self.tb.msg_connect(src, 'strobe', op, 'in')
        self.tb.start()
        # This sleep is needed since GNURadio messages are handled
        # asynchronously by the GNURadio scheduler. Unfortunately, even the
        # tests for message sinks in the gnuradio repository itself uses sleeps.
        time.sleep(0.1)
        self.tb.stop()
        self.tb.wait()

        # Check that the passed PMT blobs are accessible from starcoder_observe
        # in the expected serialized binary format.
        # self.assertEqual(op.starcoder_observe(), pmt.serialize_str(msg1))
        # self.assertEqual(op.starcoder_observe(), pmt.serialize_str(msg2))

        # Check that after retrieving all available messages, starcoder_observe
        # returns the empty string.
        # self.assertEqual(op.starcoder_observe(), '')
        # self.assertEqual(op.starcoder_observe(), '')


if __name__ == '__main__':
    gr_unittest.run(qa_enqueue_message_sink, "qa_enqueue_message_sink.xml")

File: gr-starcoder/lib/qa_enqueue_message_sink.cc
/* -*- c++ -*- */
/*
 * Copyright 2018 Infostellar, Inc.
 *
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3, or (at your option)
 * any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this software; see the file COPYING.  If not, write to
 * the Free Software Foundation, Inc., 51 Franklin Street,
 * Boston, MA 02110-1301, USA.
 */

#include "qa_enqueue_message_sink.h"
#include <cppunit/TestAssert.h>
#include <gnuradio/attributes.h>
#include <gnuradio/blocks/message_strobe.h>
#include <gnuradio/top_block.h>
#include <starcoder/enqueue_message_sink.h>
#include <stdio.h>
#include <string_queue.h>
#include <chrono>
#include <thread>

namespace gr {
namespace starcoder {

void qa_enqueue_message_sink::test_no_registered_queue() {
  gr::top_block_sptr tb = gr::make_top_block("top");
  // The following source block is needed even if we don't use its message.
  // This is because the GNURadio scheduler will not create a thread for
  // blocks that aren't connected to anything. Thus the following block is
  // needed to have enqueue_message_sink connected to something.
  gr::block_sptr src = gr::blocks::message_strobe::make(pmt::mp("in"), 1000);
  gr::starcoder::enqueue_message_sink::sptr op =
      gr::starcoder::enqueue_message_sink::make();

  gr::basic_block_sptr bb = op->to_basic_block();
  bb->_post(pmt::mp("in"), pmt::make_u8vector(10, 97));

  tb->msg_connect(src, pmt::mp("strobe"), op, pmt::mp("in"));
  tb->start();

  // This sleep is needed since GNURadio messages are handled
  // asynchronously by the GNURadio scheduler. Unfortunately, even the
  // tests for message sinks in the gnuradio repository itself uses sleeps.
  std::this_thread::sleep_for(std::chrono::milliseconds(100));

  tb->stop();
  tb->wait();
}

void qa_enqueue_message_sink::test_registered_queue() {
  gr::top_block_sptr tb = gr::make_top_block("top");
  // The following source block is needed even if we don't use its message.
  // This is because the GNURadio scheduler will not create a thread for
  // blocks that aren't connected to anything. Thus the following block is
  // needed to have enqueue_message_sink connected to something.
  gr::block_sptr src = gr::blocks::message_strobe::make(pmt::mp("in"), 1000);
  gr::starcoder::enqueue_message_sink::sptr op =
      gr::starcoder::enqueue_message_sink::make();
  string_queue q(10);

  op->register_starcoder_queue(q.get_ptr());

  gr::basic_block_sptr bb = op->to_basic_block();
  bb->_post(pmt::mp("in"), pmt::make_u8vector(10, 97));  // 97 is ascii for 'a'
  bb->_post(pmt::mp("in"), pmt::make_u8vector(3, 98));   // 98 is ascii for 'b'

  tb->msg_connect(src, pmt::mp("strobe"), op, pmt::mp("in"));
  tb->start();

  // This sleep is needed since GNURadio messages are handled
  // asynchronously by the GNURadio scheduler. Unfortunately, even the
  // tests for message sinks in the gnuradio repository itself uses sleeps.
  std::this_thread::sleep_for(std::chrono::milliseconds(100));

  tb->stop();
  tb->wait();

  // TODO (rei): Explicitly check the contents of the returned string are in the
  // expected
  // binary format. Currently I'm unable to include the protobuf class in C++
  // tests.
  CPPUNIT_ASSERT_EQUAL(q.pop().size(), (size_t) 12);
  CPPUNIT_ASSERT_EQUAL(q.pop().size(), (size_t) 5);

  // Check that after retrieving all available messages, the queue
  // returns the empty string.
  std::string empty_string = "";
  CPPUNIT_ASSERT_EQUAL(q.pop(), empty_string);
  CPPUNIT_ASSERT_EQUAL(q.pop(), empty_string);
}

} /* namespace starcoder */
} /* namespace gr */
